README.md
# Alibaba Cloud Python Software Development Kit

[![PyPI version](https://badge.fury.io/py/aliyun-python-sdk-core.svg)](https://badge.fury.io/py/aliyun-python-sdk-core)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/6c81a06478ab41f7b1e803c0d4443c73)](https://app.codacy.com/app/aliyun/aliyun-openapi-python-sdk?utm_source=github.com&utm_medium=referral&utm_content=aliyun/aliyun-openapi-python-sdk&utm_campaign=Badge_Grade_Dashboard)
[![Build Status](https://travis-ci.org/aliyun/aliyun-openapi-python-sdk.svg?branch=master)](https://travis-ci.org/aliyun/aliyun-openapi-python-sdk)
[![Build status](https://ci.appveyor.com/api/projects/status/ddq0kwqqbep9jasi/branch/master?svg=true)](https://ci.appveyor.com/project/aliyun/aliyun-openapi-python-sdk/branch/master)
[![codecov](https://codecov.io/gh/aliyun/aliyun-openapi-python-sdk/branch/master/graph/badge.svg)](https://codecov.io/gh/aliyun/aliyun-openapi-python-sdk)

[中文文档](./README_zh.md)

The Alibaba Cloud Python Software Development Kit (SDK) allows you to access Alibaba Cloud services such as Elastic Compute Service (ECS), Object Storage Service (OSS), and Resource Access Management (RAM).  You can access Alibaba Cloud services without the need to handle API related tasks, such as signing and constructing your requests.

This document introduces how to obtain and call Alibaba Cloud Python SDK.


## Online Demo

**[API Explorer](https://api.aliyun.com)** provides the ability to call the cloud product OpenAPI online, and dynamically generate SDK Example code and quick retrieval interface, which can significantly reduce the difficulty of using the cloud API. **It is highly recommended**.

<a href="https://api.aliyun.com" target="api_explorer">
  <img src="https://img.alicdn.com/tfs/TB12GX6zW6qK1RjSZFmXXX0PFXa-744-122.png" width="180" />
</a>

## Documentation

- [Requirements](docs/0-Requirement-EN.md)
- [Installation](./docs/1-Installation-EN.md)
- [Client & Credentials](./docs/2-Client-EN.md)
- [Timeout](./docs/3-Timeout-EN.md)
- [Proxy Configurations](./docs/4-Proxy-EN.md)
- [Log](./docs/5-Log-EN.md)
- [Endpoint](./docs/6-Endpoint-EN.md)

## Prerequisites

- To use Alibaba Cloud Python SDK, you must have an Alibaba Cloud account as well as an AccessKey.

	The AccessKey is required when initializing `AcsClient`. You can create an AccessKey in the Alibaba Cloud console. For more information, see [Create an AccessKey](https://usercenter.console.aliyun.com/?spm=5176.doc52740.2.3.QKZk8w#/manage/ak).

	> **Note:** To increase the security of your account, we recommend that you use the AccessKey of the RAM user to access Alibaba Cloud services.

- To use Alibaba Cloud Python SDK to access the APIs of a product, you must first activate the product on the [Alibaba Cloud console](https://home.console.aliyun.com/?spm=5176.doc52740.2.4.QKZk8w) if required.

- Alibaba Cloud Python SDK requires 2.6.x, 2.7.x, and Python 3.x.


## Install Python SDK

Alibaba Cloud Python SDK supports Python 2.6.x, 2.7.x, and Python 3.x. Run ``python --version`` to check your version of Python.

You can install the Alibaba Cloud Python SDK using the following two methods. Regardless of which method and cloud service are used, the core library `aliyun-python-sdk-core` must be installed.

- **Install with pip**

	Python SDK uses a common package management tool named `pip`. If pip is not installed, see the [pip user guide](https://pip.pypa.io/en/stable/installing/?spm=5176.doc53090.2.7.zHDiNV "pip User Guide") to install pip.

	Run the following command to install the individual libraries of Alibaba Cloud services:

	```bash
	# Install the core library
	pip install aliyun-python-sdk-core
	# Install the ECS management library
	pip install aliyun-python-sdk-ecs
	# Install the RDS management library
	pip install aliyun-python-sdk-rds
	```

## Use Python SDK

1. Import the required modules as follows:

    ```python
    from aliyunsdkcore.client import AcsClient
    from aliyunsdkcore.acs_exception.exceptions import ClientException
    from aliyunsdkcore.acs_exception.exceptions import ServerException
    from aliyunsdkecs.request.v20140526 import DescribeInstancesRequest
    from aliyunsdkecs.request.v20140526 import StopInstanceRequest
    ```
2. Initialize the `AcsClient` instance:

    ```python
    client = AcsClient(
        "<access-key-id>",
        "<access-key-secret>",
        "<region-id>"
    );
    ```

	where:

	- `access-key-id` is the Accesskey ID for your account.
	- `access-key-secret` is the AccessKey secret for your account.
	- `region-id` is the ID of the region where the service is called. For a list of region IDs, see [Regions and zones](https://www.alibabacloud.com/help/doc-detail/40654.html).

	> **Note:** The sequence of these parameters cannot be changed.

3. Initialize a request and print response.

	```python
	# Initialize a request and set parameters
	request = DescribeInstancesRequest.DescribeInstancesRequest()
	request.set_PageSize(10)
	# Print response
	response = client.do_action_with_exception(request)
	print response
	```

## Code example

The following example shows how to query a list of ECS instances in a specific region using [DescribeInstances](~~25506~~). Substitute the values for `your-access-key-id`, `your-access-key-secret`, and `your-region-id`.

```python
# -*- coding: utf8 -*-

from aliyunsdkcore.client import AcsClient
from aliyunsdkcore.acs_exception.exceptions import ClientException
from aliyunsdkcore.acs_exception.exceptions import ServerException
from aliyunsdkecs.request.v20140526 import DescribeInstancesRequest
from aliyunsdkecs.request.v20140526 import StopInstanceRequest

# Initialize AcsClient instance
client = AcsClient(
  "<your-access-key-id>",
  "<your-access-key-secret>",
  "<your-region-id>"
);

# Initialize a request and set parameters
request = DescribeInstancesRequest.DescribeInstancesRequest()
request.set_PageSize(10)

# Print response
response = client.do_action_with_exception(request)
print response
```

## HTTP DEBUG
To use the function `HTTP DEBUG`, you must set `DEBUG` in your environment variable, the corresponding value
may be `sdk` or `SDK`.

The following example shows what the `HTTP DEBUG` do, which will help you debug your codes.

```
> GET /databases?RegionId=cn-hangzhou HTTP/1.1
> Host : ads.cn-hangzhou.aliyuncs.com
> User-Agent : AlibabaCloud (Windows 10;AMD64) Python/3.7.1 Core/2.13.1 python-requests/2.18.1
> accept-encoding : *
> Accept : application/json
> Connection : keep-alive
> x-sdk-invoke-type : normal
> x-acs-version : 2019-01-22
> x-acs-region-id : cn-hangzhou
> Date : Thu, 21 Feb 2019 08:00:50 GMT
> x-acs-signature-method : HMAC-SHA1
> x-acs-signature-version : 1.0
> Authorization : acs ...
> x-sdk-client : python/2.0.0

< HTTP/1.1 503 SERVICE_UNAVAILABLE
< Date : Thu, 21 Feb 2019 08:00:50 GMT
< Content-Type : application/json; charset=UTF-8
< Content-Length : 297
< Connection : keep-alive
< Access-Control-Allow-Origin : *
< Access-Control-Allow-Methods : POST, GET, OPTIONS
< Access-Control-Allow-Headers : X-Requested-With, X-Sequence, _aop_secret, _aop_signature
< Access-Control-Max-Age : 172800
< x-acs-request-id : 670F3D09-F8E7-4144-83C3-B56C35DA35ED
< Server : Jetty(7.2.2.v20101205)
```


aliyun-python-sdk-cs/ChangeLog.txt
2020-09-23 Version: 4.8.1
- Some new parameters are supported.

2020-06-23 Version: 4.8.0
- Generated 2015-12-15 for `CS`.

2020-06-23 Version: 1.1.0
- Generated 2015-12-15 for `CS`.

2020-01-15 Version: 4.7.14
- Add Workflow csk.

2020-01-15 Version: 4.0.5
- Support Workflow csk.

2019-05-16 Version: 4.7.14
1, Update SDK Version
2, Change api UpgradeClusterAddons, UpgradeClusterComponents, DescribeClusterAddonsVersion to public


2019-05-16 Version: 4.7.13
1, Update SDK Version
2, Change api UpgradeClusterAddons, UpgradeClusterComponents, DescribeClusterAddonsVersion to public


2019-05-16 Version: 4.7.12
1, Update SDK version
2, Add api DescribeClusterAddonsVersion


2019-03-19 Version: 2.9.1
1, Update Dependency

2018-02-27 Version: 2.9.0
1, add new api DescribeClusters,DescribeClusterLogs,DescribeClusterNodes,UpgradeClusterComponents

2017-09-27 Version: 2.2.1
1, upgrade setup.py to support python3

2017-09-04 Version: 2.2.0
1, 增加了DescribeClusterScaledNodeRequest的接口。
2, 可以通过该接口查询容器集群中通过弹性节点伸缩触发器创建的ECS信息。




python-sdk-functional-test/base.py
# Copyright 2018 Alibaba Cloud Inc. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os.path
import json
import os
import threading
import logging
import time

import sys
from aliyunsdkcore.client import AcsClient
from aliyunsdkcore.vendored.six import iteritems
from aliyunsdkcore.acs_exception.exceptions import ServerException

from aliyunsdkram.request.v20150501.ListUsersRequest import ListUsersRequest
from aliyunsdkram.request.v20150501.CreateUserRequest import CreateUserRequest
from aliyunsdkram.request.v20150501.CreateAccessKeyRequest import CreateAccessKeyRequest
from aliyunsdkram.request.v20150501.DeleteAccessKeyRequest import DeleteAccessKeyRequest
from aliyunsdkram.request.v20150501.ListAccessKeysRequest import ListAccessKeysRequest
from aliyunsdkram.request.v20150501.ListRolesRequest import ListRolesRequest
from aliyunsdkram.request.v20150501.CreateRoleRequest import CreateRoleRequest
from aliyunsdkram.request.v20150501.DeleteRoleRequest import DeleteRoleRequest
from aliyunsdkram.request.v20150501.AttachPolicyToUserRequest import AttachPolicyToUserRequest

# The unittest module got a significant overhaul
# in 2.7, so if we're in 2.6 we can use the backported
# version unittest2.
if sys.version_info[:2] == (2, 6):
    from unittest2 import TestCase
else:
    from unittest import TestCase

# the version under py3 use the different package
if sys.version_info[0] == 3:
    from http.server import SimpleHTTPRequestHandler
    from http.server import HTTPServer
else:
    from SimpleHTTPServer import SimpleHTTPRequestHandler
    from BaseHTTPServer import HTTPServer


def request_helper(client, request, **params):
    for key, value in iteritems(params):
        set_name = 'set_' + key
        if hasattr(request, set_name):
            func = getattr(request, set_name)
            func(value)
        else:
            raise Exception(
                "{0} has no parameter named {1}.".format(request.__class__.__name__, key))
    response = client.do_action_with_exception(request)
    return json.loads(response.decode('utf-8'))


def _check_server_response(obj, key):
    if key not in obj:
        raise Exception("No '{0}' in server response.".format(key))


def find_in_response(response, key=None, keys=None):
    if key:
        _check_server_response(response, key)
        return response[key]
    if keys:
        obj = response
        for key in keys:
            _check_server_response(obj, key)
            obj = obj[key]
        return obj


class SDKTestBase(TestCase):

    def __init__(self, *args, **kwargs):
        TestCase.__init__(self, *args, **kwargs)
        # if sys.version_info[0] == 2:
        #     self.assertRegex = self.assertRegexpMatches
        self._init_env()

    def test_env_available(self):
        # To let test script know whether env is available, to continue the tests
        self._init_env()

    def _init_env(self):
        self._sdk_config = self._init_sdk_config()
        self.access_key_id = self._read_key_from_env_or_config("ACCESS_KEY_ID")
        self.access_key_secret = self._read_key_from_env_or_config("ACCESS_KEY_SECRET")
        self.region_id = self._read_key_from_env_or_config("REGION_ID")
        self.user_id = self._read_key_from_env_or_config("USER_ID")
        self.root_user_id = self._read_key_from_env_or_config("ROOT_UID")
        if 'TRAVIS_JOB_NUMBER' in os.environ:
            self.travis_concurrent = os.environ.get('TRAVIS_JOB_NUMBER').split(".")[-1]
        else:
            self.travis_concurrent = "0"
        self.default_ram_user_name = "RamUserForSDKCredentialsTest" + self.travis_concurrent
        self.default_ram_role_name = "RamROleForSDKTest" + self.travis_concurrent
        self.default_role_session_name = "RoleSession" + self.travis_concurrent
        self.ram_user_id = None
        self.ram_policy_attched = False
        self.ram_user_access_key_id = None
        self.ram_user_access_key_secret = None
        self.ram_role_arn = None

    def _init_sdk_config(self):
        sdk_config_path = os.path.join(os.path.expanduser("~"), "aliyun_sdk_config.json")
        if os.path.isfile(sdk_config_path):
            with open(sdk_config_path) as fp:
                return json.loads(fp.read())

    def _read_key_from_env_or_config(self, key_name):
        if key_name.upper() in os.environ:
            return os.environ.get(key_name.upper())
        if key_name.lower() in self._sdk_config:
            return self._sdk_config[key_name.lower()]

        raise Exception("Failed to find sdk config: " + key_name)

    def setUp(self):
        TestCase.setUp(self)
        self.client = self.init_client()

    def tearDown(self):
        pass

    def init_client(self, region_id=None):
        if not region_id:
            region_id = self.region_id
        client = AcsClient(self.access_key_id, self.access_key_secret, region_id, timeout=120)
        client.set_stream_logger()
        return client

    @staticmethod
    def get_dict_response(string):
        return json.loads(string.decode('utf-8'), encoding="utf-8")

    def _create_default_ram_user(self):
        if self.ram_user_id:
            return
        response = request_helper(self.client, ListUsersRequest())
        user_list = find_in_response(response, keys=['Users', 'User'])
        for user in user_list:
            if user['UserName'] == self.default_ram_user_name:
                self.ram_user_id = user["UserId"]
                return

        response = request_helper(self.client, CreateUserRequest(),
                                  UserName=self.default_ram_user_name)
        self.ram_user_id = find_in_response(response, keys=['User', 'UserId'])

    def _attach_default_policy(self):
        if self.ram_policy_attched:
            return

        try:
            request_helper(self.client, AttachPolicyToUserRequest(),
                           PolicyType='System', PolicyName='AliyunSTSAssumeRoleAccess',
                           UserName=self.default_ram_user_name)
        except ServerException as e:
            if e.get_error_code() == 'EntityAlreadyExists.User.Policy':
                pass
            else:
                raise e

        self.ram_policy_attched = True

    def _create_access_key(self):
        if self.ram_user_access_key_id and self.ram_user_access_key_secret:
            return

        response = request_helper(self.client, ListAccessKeysRequest(),
                                  UserName=self.default_ram_user_name)
        for access_key in find_in_response(response, keys=['AccessKeys', 'AccessKey']):
            access_key_id = access_key['AccessKeyId']
            request_helper(self.client, DeleteAccessKeyRequest(),
                           UserAccessKeyId=access_key_id,
                           UserName=self.default_ram_user_name)

        response = request_helper(self.client, CreateAccessKeyRequest(),
                                  UserName=self.default_ram_user_name)
        self.ram_user_access_key_id = find_in_response(response, keys=['AccessKey', 'AccessKeyId'])
        self.ram_user_access_key_secret = find_in_response(
            response,
            keys=['AccessKey', 'AccessKeySecret'])

    def _delete_access_key(self):
        request_helper(self.client, DeleteAccessKeyRequest(),
                       UserName=self.default_ram_user_name,
                       UserAccessKeyId=self.ram_user_access_key_id)

    def init_sub_client(self):
        self._create_default_ram_user()
        self._attach_default_policy()
        self._create_access_key()
        client = AcsClient(self.ram_user_access_key_id,
                           self.ram_user_access_key_secret,
                           self.region_id, timeout=120)
        return client

    def _create_default_ram_role(self):
        if self.ram_role_arn:
            return
        response = request_helper(self.client, ListRolesRequest())
        for role in find_in_response(response, keys=['Roles', 'Role']):
            role_name = role['RoleName']
            role_arn = role['Arn']
            if role_name == self.default_ram_role_name:
                self.ram_role_arn = role_arn
                return

        policy_doc = """
        {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "RAM": [
                  "acs:ram::%s:root"
                ]
              }
            }
          ],
          "Version": "1"
        }
        """ % self.root_user_id

        response = request_helper(self.client, CreateRoleRequest(),
                                  RoleName=self.default_ram_role_name,
                                  AssumeRolePolicyDocument=policy_doc)
        self.ram_role_arn = find_in_response(response, keys=['Role', 'Arn'])
        # FIXME We have wait for 5 seconds after CreateRole before
        # we can AssumeRole later
        time.sleep(5)

    def _delete_default_ram_role(self):
        request_helper(self.client, DeleteRoleRequest(), RoleName=self.default_ram_role_name)


def disabled(func):
    def _decorator(func):
        pass

    return _decorator


class MyServer:
    _headers = {}
    _url = ''

    def __enter__(self):
        class ServerHandler(SimpleHTTPRequestHandler):

            def do_GET(_self):
                _self.protocol_version = 'HTTP/1.1'
                self._headers = _self.headers
                self._url = _self.path
                _self.send_response(200)
                _self.send_header("Content-type", "application/json")
                _self.end_headers()
                _self.wfile.write(b"{}")

        self.server = HTTPServer(("", 51352), ServerHandler)

        def thread_func():
            self.server.serve_forever()

        thread = threading.Thread(target=thread_func)
        thread.start()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        if self.server:
            self.server.shutdown()
            self.server = None

    @property
    def headers(self):
        return self._headers

    @property
    def url(self):
        return self._url
